### PR TITLE
Migrate pxf-hbase to Spring Boot

### PR DESCRIPTION
**Pull request by frankgh**
- Update top level build.gradle to include pxf-hbase
- Add new build.gradle for pxf-hbase
- all existing unit tests should be ported to junit 5
- all automation tests for HBase should be working

In this PR, we intentionally comment out the automation tests for Hive, which should be restored with the inclusion of the `pxf-hive` profile.

### CODE REVIEW

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/MultipleHiveFragmentsPerFileFragmenter.java [hunk 1]
```diff
@@ -1,122 +1,122 @@
...
+//
+//        for (InputSplit split : splits) {
+//            FileSplit fsp = (FileSplit) split;
+//            String[] hosts = fsp.getLocations();
+//            String filepath = fsp.getPath().toString();
+//            return filepath;
+//        }
+//        throw new RuntimeException("Unable to get file path for table.");
+//    }
+//}
```
Comment (denalex) [inline automation/src/main/java/org/greenplum/pxf/automation/testplugin/MultipleHiveFragmentsPerFileFragmenter.java:122]: when will this be uncommented ?

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveInputFormatFragmenterWithFilter.java [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
+//        LOG.debug("user defined filter: " + filterStr);
+//        if ((filterStr == null) || filterStr.isEmpty() || "null".equals(filterStr))
+//            return;
+//
+//            context.setFilterString(filterStr);
+//            LOG.debug("User defined filter: " + context.getFilterString());
+//
+//            LOG.debug("User defined filter: " + context.hasFilter());
+//    }
+//}
```
Comment (denalex) [inline automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveInputFormatFragmenterWithFilter.java:35]: when will this be uncommented ?

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveDataFragmenterWithFilter.java [hunk 3]
```diff
@@ -1,35 +1,35 @@
...
+//        LOG.debug("user defined filter: " + filterStr);
+//        if ((filterStr == null) || filterStr.isEmpty() || "null".equals(filterStr))
+//            return;
+//
+//        context.setFilterString(filterStr);
+//        LOG.debug("User defined filter: " + context.getFilterString());
+//
+//        LOG.debug("User defined filter: " + context.hasFilter());
+//    }
+//}
```
Comment (denalex) [inline automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveDataFragmenterWithFilter.java:35]: when will this be uncommented ?

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/FilterVerifyFragmentMetadata.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.greenplum.pxf.automation.testplugin;
+
+import org.greenplum.pxf.api.utilities.FragmentMetadata;
+
+public class FilterVerifyFragmentMetadata implements FragmentMetadata {
```
Comment (denalex) [inline automation/src/main/java/org/greenplum/pxf/automation/testplugin/FilterVerifyFragmentMetadata.java:5]: this does not need @JsonCreator because it has a single argument constructor and there's only 1 property ?

File: server/pxf-hbase/build.gradle [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+     * Implementation Dependencies
+     *******************************/
+
+    implementation(project(':pxf-api'))
+    implementation("org.apache.hbase:hbase-client:${hbaseVersion}")
+    implementation("org.apache.hbase:hbase-common:${hbaseVersion}")
+    implementation("org.apache.hbase:hbase-protocol:${hbaseVersion}")
+    implementation("org.apache.htrace:htrace-core:3.1.0-incubating")
+    implementation("io.netty:netty-all:4.0.23.Final")
+    implementation("org.apache.zookeeper:zookeeper:3.4.6")
```
Comment (denalex) [inline server/pxf-hbase/build.gradle:22]: wonder why we need the whole netty and the whole zookeeper (where only client jars would suffice) ? also is yammer still needed with Boot ?

File: server/pxf-hbase/src/main/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadata.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    @Getter
+    private final Map<String, byte[]> columnMapping;
+
+    public HBaseFragmentMetadata(HRegionInfo region, Map<String, byte[]> columnMapping) {
+        this(region.getStartKey(), region.getEndKey(), columnMapping);
+    }
+
+    @JsonCreator
+    public HBaseFragmentMetadata(
+            @JsonProperty("startKey") byte[] startKey,
```
Comment (denalex) [inline server/pxf-hbase/src/main/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadata.java:28]: wonder how serialized byte[] looks like in JSON ? Is it a base64 encoded string ?

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseAccessorTest.java [hunk 7]
```diff
@@ -73,47 +67,44 @@ public void tearDown() throws Exception {
...
+     */
     @Test
-    public void construction() throws Exception {
+    public void construction() {
         prepareConstruction();
         HBaseAccessor accessor = new HBaseAccessor();
-        accessor.initialize(context);
-        PowerMockito.verifyNew(HBaseTupleDescription.class).withArguments(context);
+        accessor.setRequestContext(context);
+        accessor.initialize();
```
Comment (denalex) [inline server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseAccessorTest.java:84]: the more I see this, the more it reminds me of https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/InitializingBean.html -- wonder if we should take out explicit initialization in plugins and have them implement `afterPropertiesSet()` instead and have it being managed by Spring and not the Bridge. The tests will still have to do it, though, like here. Just food for thought.

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseAccessorTest.java [hunk 8]
```diff
@@ -156,11 +142,10 @@ private void prepareTableOpen() throws Exception {
      * Helper for test setup.
      * Sets zero columns (not realistic) and no filter
      */
-    private void prepareEmptyScanner() throws Exception {
+    private void prepareEmptyScanner() {
         scanDetails = mock(Scan.class);
-        PowerMockito.whenNew(Scan.class).withNoArguments().thenReturn(scanDetails);
 
-        when(tupleDescription.columns()).thenReturn(0);
+        // when(tupleDescription.columns()).thenReturn(0);
```
Comment (denalex) [inline server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseAccessorTest.java:148]: remove ?

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFilterBuilderTest.java [hunk 9]
```diff
@@ -94,12 +90,11 @@ public void parseNotExpressionIgnored() throws Exception {
 
     @Test
     public void parseNotOpCodeInConstant() throws Exception {
-        thrown.expect(NullPointerException.class);
 
         String filter = "a1c25s2dl2o1a1c20s1d2o2l0";
         // Testing that we get past the parsing stage
         // Very crude but it avoids instantiating all the necessary dependencies
-        assertNull(helper(filter, null));
+        assertThrows(NullPointerException.class, () -> helper(filter, null));
```
Comment (denalex) [inline server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFilterBuilderTest.java:97]: did behavior switch from returning `null` to throwing NPE ? Is it Ok ?

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadataTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        columnMapping.put("foo", fooValue);
+
+        HBaseFragmentMetadata metadata = new HBaseFragmentMetadata(startKey, endKey, columnMapping);
+        assertNotNull(metadata);
+        assertSame(startKey, metadata.getStartKey());
+        assertSame(endKey, metadata.getEndKey());
+        assertSame(columnMapping, metadata.getColumnMapping());
+        assertSame(fooValue, metadata.getColumnMapping().get("foo"));
+    }
+
```
Comment (denalex) [inline server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadataTest.java:51]: maybe also have a few tests to serialize / deserialize the POJO with ObjectMapper to make sure @JsonCreator and others are configured correctly ?

File: server/pxf-jdbc/build.gradle [hunk 11]
```diff
@@ -9,12 +9,11 @@ dependencies {
      * API Dependencies
      *******************************/
 
-    api(project(':pxf-api'))
-
     /*******************************
      * Implementation Dependencies
      *******************************/
 
+    implementation(project(':pxf-api'))
```
Comment (denalex) [inline server/pxf-jdbc/build.gradle:16]: wonder why the change

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/MultipleHiveFragmentsPerFileFragmenter.java (hunk 1 above)
Comment (frankgh) [reply automation/src/main/java/org/greenplum/pxf/automation/testplugin/MultipleHiveFragmentsPerFileFragmenter.java:122]: In the PR message I state that `we intentionally comment out the automation tests for Hive, which should be restored with the inclusion of the pxf-hive profile.` . So with the pxf-hive profile, these classes have to be uncommented

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/FilterVerifyFragmentMetadata.java (hunk 4 above)
Comment (frankgh) [reply automation/src/main/java/org/greenplum/pxf/automation/testplugin/FilterVerifyFragmentMetadata.java:5]: correct

File: server/pxf-jdbc/build.gradle (hunk 11 above)
Comment (frankgh) [reply server/pxf-jdbc/build.gradle:16]: api is not an api dependency for pxf-jdbc

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFilterBuilderTest.java (hunk 9 above)
Comment (frankgh) [reply server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFilterBuilderTest.java:97]: this one threw me off too, if you look closely, this class expects a NPE in line 97 of the previous version of this class.

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseAccessorTest.java (hunk 7 above)
Comment (frankgh) [reply server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseAccessorTest.java:84]: interesting. I did not know about it.

File: server/pxf-hbase/src/main/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadata.java (hunk 6 above)
Comment (frankgh) [reply server/pxf-hbase/src/main/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadata.java:28]: i will add some tests to see what the output looks like

File: server/pxf-hbase/build.gradle (hunk 5 above)
Comment (frankgh) [reply server/pxf-hbase/build.gradle:22]: not sure, I can comment these out and see if it runs correctly

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadataTest.java (hunk 10 above)
Comment (frankgh) [reply server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadataTest.java:51]: we have tests for json serialization / deserialization in `FragmentMetadataSerDeTest`

Comment (denalex) [reply server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadataTest.java:51]: yes, but SerDe is specific to the POJO, let's say if the names of properties on `@JsonCreator` are mis-types or mixed up, this specific POJO will have problems deserializing, while others will be ok. The test will essentially ensure that the annotation is configured correctly

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveDataFragmenterWithFilter.java (hunk 3 above)
Comment (frankgh) [reply automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveDataFragmenterWithFilter.java:35]: it will be uncommented with pxf-hive

File: automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveInputFormatFragmenterWithFilter.java (hunk 2 above)
Comment (frankgh) [reply automation/src/main/java/org/greenplum/pxf/automation/testplugin/HiveInputFormatFragmenterWithFilter.java:35]: it will be uncommented with pxf-hive

File: server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadataTest.java (hunk 10 above)
Comment (frankgh) [reply server/pxf-hbase/src/test/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadataTest.java:51]: true, I am not too keen on `@JsonCreator`. I much prefer clean POJOs with no annotations, but I will add tests to serialize, deserialize this class specifically.

File: server/pxf-hbase/src/main/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadata.java (hunk 6 above)
Comment (frankgh) [reply server/pxf-hbase/src/main/java/org/greenplum/pxf/plugins/hbase/HBaseFragmentMetadata.java:28]: just checked (and added a test), it is a base64 encoded string
